### PR TITLE
Allow setting the RPORT option for pipe_auditor

### PR DESCRIPTION
**Pull request by NtAlexio2**
This PR adds the same pattern in #19163 to `pipe_auditor` and also fixes a bug in `smb_enumusers`, which causes invalid operation in case **Empty/Zero RPORT**.

### CODE REVIEW

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
Changes look pretty good to me. I did notice however that the reporting block on lines 77-88 are not reporting the correct port now that we're handling multiple cases.

Also `@rport` will need to be set on L44 as well so it's populated in the reported note. Something like `@rport = datastore['RPORT'] = session.port` would work.

File: modules/auxiliary/scanner/smb/pipe_auditor.rb [hunk 1]
```diff
@@ -52,12 +68,12 @@ def run_host(ip)
           disconnect
           break
```
[the remark this answers is not in the document]
Comment (NtAlexio2) [reply modules/auxiliary/scanner/smb/pipe_auditor.rb]: applied ;)

File: modules/auxiliary/scanner/smb/pipe_auditor.rb [hunk 2]
```diff
@@ -35,29 +41,50 @@ def run_host(ip)
     if session
       print_status("Using existing session #{session.sid}")
       client = session.client
-      datastore['RPORT'] = session.port
+      @rport = datastore['RPORT'] = session.port
       self.simple = ::Rex::Proto::SMB::SimpleClient.new(client.dispatcher.tcp_socket, client: client)
       self.simple.connect("\\\\#{session.address}\\IPC$")
       pipes += check_pipes
```
Comment (smcintyre-r7) [inline modules/auxiliary/scanner/smb/pipe_auditor.rb:47]:
```suggestion
      report_pipes(ip, check_pipes)
```

File: modules/auxiliary/scanner/smb/pipe_auditor.rb [hunk 3]
```diff
@@ -35,29 +41,50 @@ def run_host(ip)
...
+        @rport = smb_service[:port]
+        @smb_direct = smb_service[:direct]
 
         begin
           connect
           smb_login
           pipes += check_pipes
           disconnect
-          break
+          report_pipes(pipes)
```
Comment (smcintyre-r7) [inline modules/auxiliary/scanner/smb/pipe_auditor.rb:69]:
```suggestion
          report_pipes(ip, pipes)
```

File: modules/auxiliary/scanner/smb/pipe_auditor.rb [hunk 4]
```diff
@@ -35,29 +41,50 @@ def run_host(ip)
...
 
+  def check_pipes
+    pipes = []
+    check_named_pipes.each do |pipe_name, _|
+      pipes.push(pipe_name)
+    end
+    pipes
+  end
+
+  def report_pipes(pipes)
```
Comment (smcintyre-r7) [inline modules/auxiliary/scanner/smb/pipe_auditor.rb:87]:
```suggestion
  def report_pipes(ip, pipes)
```

**Review by smcintyre-r7 — CHANGES_REQUESTED** [review]
I found the following two issues:

1. The pipes aren't reported when using a session.
2. The `#report_pipes` function fails because `ip` isn't defined. It needs to be passed as a parameter.

**Comment by NtAlexio2** [comment]
Thanks very much! updated

**Review by smcintyre-r7 — APPROVED** [review]
Changes look good to me now. I was able to test it with a session and without and everything is being reported properly. Thanks for updating this!

```
msf6 auxiliary(scanner/smb/pipe_auditor) > run SESSION=-1

[*] Using existing session 1
[+] Pipes: \netlogon, \lsarpc, \samr, \atsvc, \epmapper, \eventlog, \InitShutdown, \lsass, \LSM_API_service, \ntsvcs, \protected_storage, \router, \scerpc, \srvsvc, \W32TIME_ALT, \wkssvc
[*] Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/smb/pipe_auditor) > notes

Notes
=====

 Time                     Host            Service  Port  Protocol  Type         Data
 ----                     ----            -------  ----  --------  ----         ----
 2024-09-26 13:17:32 UTC  192.168.159.10  smb      445   tcp       Pipes Found  "Pipes: \\netlogon, \\lsarpc, \\samr, \\atsvc, \\epmapper, \\eventlog, \\InitShutdown, \\lsass, \\LSM_API_service, \\ntsvcs, \\protected_storage, \\router, \\scerpc, \\srvsvc, \\W32TIME_ALT, \\wkssvc"

msf6 auxiliary(scanner/smb/pipe_auditor) > notes -d

Notes
=====

 Time                     Host            Service  Port  Protocol  Type         Data
 ----                     ----            -------  ----  --------  ----         ----
 2024-09-26 13:17:32 UTC  192.168.159.10  smb      445   tcp       Pipes Found  "Pipes: \\netlogon, \\lsarpc, \\samr, \\atsvc, \\epmapper, \\eventlog, \\InitShutdown, \\lsass, \\LSM_API_service, \\ntsvcs, \\protected_storage, \\router, \\scerpc, \\srvsvc, \\W32TIME_ALT, \\wkssvc"

[*] Deleted 1 notes
msf6 auxiliary(scanner/smb/pipe_auditor) > run

[+] 192.168.159.10:445 - Pipes: \netlogon, \lsarpc, \samr
[*] 192.168.159.10: - Scanned 1 of 1 hosts (100% complete)
[*] Auxiliary module execution completed
msf6 auxiliary(scanner/smb/pipe_auditor) > notes

Notes
=====

 Time                     Host            Service  Port  Protocol  Type         Data
 ----                     ----            -------  ----  --------  ----         ----
 2024-09-26 13:18:33 UTC  192.168.159.10  smb      445   tcp       Pipes Found  "Pipes: \\netlogon, \\lsarpc, \\samr"

msf6 auxiliary(scanner/smb/pipe_auditor) > 
``

**Comment by smcintyre-r7** [comment]
# Release Notes
This refactors the `pipe_auditor` scanner module to allow an RPORT argument to be specified and refactors it to follow more recent patterns used by SMB modules.

**Pull request merged**